docs/src/tutorials/algorithms/tsp_lazy_constraints.jl

```diff
@@ -0,0 +1,326 @@
+
+seen = Dict()
+
+function subtour_elimination(cb_data)
+    ## We only checkfor subtours when we encounter integer-feasible solutions
+    status = callback_node_status(cb_data, tsp)
+    if status == MOI.CALLBACK_NODE_STATUS_INTEGER
+
+        ## Load the callback data at the current node  
+        x_val = callback_value.(Ref(cb_data), x)
+
+        ## Write the current edges in a tuple list and identify the shortest  cycle
+        edges = selected_edges(x_val)
+        cycle = subtour(edges)
+        h = hash(sort(cycle))
+
+        if !(h in keys(seen))
+            ## A subtour contains at least 2 locations and at most (n-1)
+            if length(cycle) > 1 && length(cycle) < n
+                S = collect(
+                    Iterators.filter(
+                        t -> t[1] < t[2],
+                        Iterators.product(cycle, cycle),
+                    ),
+                )
+                con = @build_constraint(
+                    sum(x[t[1], t[2]] for t in S) <= length(cycle) - 1
+                )
+
+                MOI.submit(tsp, MOI.LazyConstraint(cb_data), con)
+                seen[h] = con
+            end
+        end
+    end
+end
+
+# For solving the model with lazy constraints, we need to make sure that `LazyConstraints` are enabled and that the proper `CallbackFunction()` passed.
+# Before we do this, we are going to quickly rebuild the model, such that previously added subtour elimination constraints are removed.
+
+tsp = Model(GLPK.Optimizer);
+set_optimizer_attribute(tsp, "sr_heur", GLPK.GLP_OFF)
+## set_optimizer_attribute(tsp, "msg_lev", GLPK.GLP_MSG_ON)
+
+@variable(tsp, x[i in 1:n, j in 1:n], Bin);
+
+@objective(tsp, Min, sum(x[i, j] * d[i, j] / 2 for i in 1:n for j in 1:n));
+
+@constraint(tsp, symmetry[i in 1:n, j in i:n, i!=j], x[i, j] == x[j, i]);
+
+@constraint(tsp, two_degree[i in 1:n], sum(x[i, j] for j in 1:n) == 2);
+
+@constraint(tsp, no_loop[i in 1:n], x[i, i] == 0);
+
+MOI.set(tsp, MOI.LazyConstraintCallback(), subtour_elimination)
+optimize!(tsp)
+time_lazy = solve_time(tsp);
+
+# !!! warning
+#     The following is a bit of a hack, as different solvers appear to treat lazy constraints internally very differently!
+#     While the following part is not necessary when using `Gurobi` as an Optimizer, in the case of `GLPK` things are different:
+#     in combination with the dict `lazy_constraints` introduced above, some lazy constraints may be ignored, even though they were explicitely added to the model.
+#     This might be remedied by removing the dict; however, then the optimization using lazy constraints will be very slow.
+#     Therefore, for use with `GLPK`,  after the first solution we will verify that all previously identified lazy constraints are actually respected, by adding them to the model and re-solving it.
+#     Note that this fix also works for the `sr_heur` issue discussed at the beginning of this document.
```

Comment:
Hmm. Perhaps I need to look into this in more detail. There might be something we can do in the callback.

Reply:
Thanks, I would be grateful for any further suggestions or assistance. I agree with your initial comment that it feels unpleasant to end the tutorial with a bit of a hack!

Unfortunately, I am not very familiar with `GLPK` as an Optimizer. That is why I only realized quite late during my pull request that the lazy constraints might not be working as expected when using that Optimizer. You can easily verify for yourself that the final solution would incorrect (for the fixed n and seed), even though some of the remaining subtours where explicitly eliminated through the callback. Just for your reference: Note that the code works as expected when using `Gurobi` as Optimizer which is why I assume that the issue might be related to `GLPK`, not `JuMP`.